sysml.library/Kernel Library/Occurrences.kerml

```diff
@@ -1,297 +1,561 @@
+	private import Base::DataValue;
+	private import ScalarValues::Integer;
+	private import ScalarValues::Boolean;
+	private import Links::*;
+	private import Collections::Set;
+	private import Collections::OrderedSet;
+	private import SequenceFunctions::includes;
+	private import CollectionFunctions::contains;
+	
+	/**
+	 * Occurrence is the most general classifier of entities that have identity and 
+	 * occur over time and space.
+	 * 
+	 * The features of Occurrence specify the semantics of associations between occurrences that
+	 * assert complete inclusion and exclusion in time or space, or both, which includes
+	 * portions of an occurrence (having the same identity).  Portions include slices and shots
+	 * over time and space.
+	 */
+	abstract class Occurrence specializes Anything {
+		private import SequenceFunctions::*;
+		
+		feature portionOfLife: Life[1] subsets portionOf;
+		
+		feature self: Occurrence[1] redefines Anything::self subsets timeSlices, spaceSlices, sameLifeOccurrences;
+		feature sameLifeOccurrences: Occurrence[1..*] subsets things;
+		
+		/**
+		 * Occurrences that end before this occurrence starts.
+		 */
+		feature predecessors: Occurrence[0..*] subsets occurrences;
+		
+		/**
+		 * Occurrences that start after this occcurrence ends.
+		 */
+		feature successors: Occurrence[0..*] subsets occurrences;
+		
+		/**
+		 * Occurrences that end just before this occurrence starts, with no
+		 * possibility of other occurrences happening in the time between them.
+		 */
+		feature immediatePredecessors: Occurrence[0..*] subsets predecessors;
+
+		/**
+		 * Occurrences that start just after this occurrence ends, with no
+		 * possibility of other occurrences happening in the time between them.
+		 */
+		feature immediateSuccessors: Occurrence[0..*] subsets successors;
+
+		/**
+		 * Occurrences that start no earlier than and end no later than
+		 * this occurrence, including at least this occurrence.
+		 */
+		feature timeEnclosedOccurrences: Occurrence[1..*] subsets occurrences;
+
+		/**
+		 * Occurrences that this one completely includes in space (not necessarily in time),
+		 * including this one.
+		 */
+		feature spaceEnclosedOccurrences: Occurrence[1..*] subsets occurrences;
+
+		/**
+		 * Occurrences that this one completely includes in both space and time,
+		 * including this one.
+		 */
+		feature spaceTimeEnclosedOccurrences: Occurrence[1..*] subsets timeEnclosedOccurrences, spaceEnclosedOccurrences;
+
+		/**
+		 * All space time enclosed occurrences that take up zero time and space.
+		 */
+		feature all spaceTimeEnclosedPoints : Occurrence[1..*] subsets spaceTimeEnclosedOccurrences {
+			redefines innerSpaceDimension = 0;
+			binding startShot[1] = endShot[1];
+		}
+
+		/**
+		 * The number of variables need to identify space points in this occurrence, from 0
+		 * to 3, without regard to higher dimensional spaces it might be emedded in.
+		 */
+		feature innerSpaceDimension : Integer [1];

```

Comment:
Can be typed `Natural[1]` as dimensions should never be negative.

Reply:
Done.